src/main/java/com/google/devtools/build/lib/remote/CasDigestLookup.java

```diff
@@ -0,0 +1,114 @@
+package com.google.devtools.build.lib.remote;
+
+import build.bazel.remote.execution.v2.ContentAddressableStorageGrpc;
+import build.bazel.remote.execution.v2.Digest;
+import build.bazel.remote.execution.v2.FindMissingBlobsRequest;
+import build.bazel.remote.execution.v2.FindMissingBlobsResponse;
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Iterables;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.MoreExecutors;
+import com.google.devtools.build.lib.remote.util.DigestUtil;
+import com.google.devtools.build.lib.remote.util.TracingMetadataUtils;
+import io.grpc.CallCredentials;
+import io.grpc.Context;
+import io.grpc.StatusRuntimeException;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.TimeUnit;
+
+public class CasDigestLookup {
```

Comment:
Why have this class? Why no reuse getMissingDigests from GrpcRemoteCache?

Reply:
From my interpretation there is a clear boundary between the GrpcRemoteCache and the ByteStreamBuildEventArtifactUploader and I didn't want to have the latter be aware of the former. Perhaps having the GrpcRemoteCache implement a CasDigestLookup interface would be a better approach, I'll do that.